curvlinops/fisher.py

```diff
@@ -1,16 +1,212 @@
+    model's predictive distribution and computing the expected outer product of
+    gradients w.r.t. the sampled targets.
+
+    The implementation works by:
+    1. Computing model predictions :math:`f_n = f_{\mathbf{\theta}}(\mathbf{x}_n)`
+    2. Flattening outputs to 2D format for loss computation:
+       - CrossEntropyLoss: ``(batch, c, ...) -> (batch*..., c)``
+       - Other losses: ``(batch, ..., c) -> (batch*..., c)``
+    3. Sampling gradients :math:`g_{nm} = \nabla_{f_n} \ell(f_n, \tilde{y}_{nm})`
+       where :math:`\tilde{y}_{nm} \sim q(\cdot | f_n)` are sampled targets
+    4. Computing a pseudo-loss :math:`L' = \frac{1}{2Mc} \sum_n \sum_m \langle f_n, g_{nm} \rangle^2`
+       where :math:`M` is the number of MC samples and :math:`c` is the reduction factor
+    5. Using the GGN of this pseudo-loss to approximate the MC Fisher matrix-vector products
+
+    The reduction factor :math:`c` adjusts for the loss function's reduction:
+    - ``'mean'``: :math:`c = N` (CrossEntropyLoss) or :math:`c = N \times C` (others)
+    - ``'sum'``: :math:`c = 1`
+
+    Args:
+        model_func: The neural network :math:`f_{\mathbf{\theta}}`.
+        loss_func: The loss function :math:`\ell`.
+        params: A tuple of parameters w.r.t. which the MC Fisher is computed.
+            All parameters must be part of ``model_func.parameters()``.
+
+    Returns:
+        A function that takes inputs ``X``, ``y``, a matrix ``M`` in list format,
+        ``mc_samples``, and ``generator``, and returns the mini-batch MC Fisher
+        applied to ``M`` in list format.
+    """
+    f_flat, _ = make_functional_flattened_model_and_loss(model_func, loss_func, params)
+
+    # Create the gradient output sampler for this loss function
+    sample_grad_output_flat = make_grad_output_sampler(loss_func)
+
+    def c_pseudo_flat(
+        output_flat: Tensor, y: Tensor, mc_samples: int, generator: Generator
+    ) -> Tensor:
+        """Compute MC-Fisher pseudo-loss: L' = 0.5 / (M * c) * sum_n sum_m <f_n, g_nm>^2.
+
+        This pseudo-loss L' := 0.5 / (M * c) ∑ₙ ∑ₘ fₙᵀ (gₙₘ gₙₘᵀ) fₙ where
+        gₙₘ = ∂ℓₙ(yₙₘ)/∂fₙ (detached) and M is the number of MC samples.
+        The GGN of L' linearized at fₙ is the MC Fisher.
+        We can thus multiply with the MC Fisher by computing the GGN-vector products of L'.
+
+        The reduction factor adjusts the scale depending on the loss reduction used.
+
+        Args:
+            output_flat: The flattened model predictions.
+            y: The un-flattened labels.
+            mc_samples: Number of MC samples to use.
+            generator: Random generator for sampling.
+
+        Returns:
+            The pseudo-loss whose GGN is the MC-Fisher.
+        """
+        # Sample gradients w.r.t. output using the provided generator
+        grad_output_samples = sample_grad_output_flat(
+            output_flat.detach(), mc_samples, y, generator
+        )
```

Comment:
This function is identical to `c_pseudo_flat` for the EF, besides the sampling here (and the adjustment of the `reduction_factor` by multiplying with `mc_samples`). Might be worth it to avoid this code duplication if possible.

Reply:
Ah I missed this, thanks! I think I actually prefer efficiency here then. If you really want you could add a comment about this.